packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/props/cache.rs

```diff
@@ -0,0 +1,254 @@
+//! Allow for the caching of props (and state props and strings).
+
+use std::{
+    borrow,
+    cell::{Ref, RefCell},
+    rc::Rc,
+};
+
+use crate::{
+    components::prelude::{GraphNode, PropCalcResult, PropValue},
+    graph::directed_graph::Taggable,
+    new_core::graph_node::GraphNodeLookup,
+    state::PropStatus,
+};
+
+/// Metadata stored along with a cached prop.
+#[derive(Debug, Clone)]
+struct CachedPropMeta {
+    /// Status of the prop indicating its initialization state and whether it is stale
+    pub status: PropStatus,
+    /// True if the value was set using the variable's default value.
+    /// Is typically used to lower the precedence of this prop's value
+    /// when calculating another prop, as it indicates that the document author
+    /// did not explicitly set the value
+    pub came_from_default: bool,
+    /// A counter that is incremented every time the value is set.
+    /// Used to compare with another counter to determine if the variable has been changed
+    /// since the other counter was last set.
+    pub change_counter: u32,
+}
+
+/// Store a prop's and other meta information needed to operate a cache
+/// that intelligently recomputes props when their dependencies change.
+#[derive(Debug)]
+struct CachedProp {
+    /// Cached value of the prop. Will be `None` if the prop hasn't been computed or is stale.
+    value: Rc<RefCell<Option<PropValue>>>,
+    meta: RefCell<CachedPropMeta>,
+}
+
+impl CachedProp {
+    pub fn new() -> Self {
+        CachedProp {
+            value: Rc::new(RefCell::new(None)),
+            meta: RefCell::new(CachedPropMeta {
+                status: PropStatus::Unresolved,
+                came_from_default: false,
+                change_counter: 0,
+            }),
+        }
+    }
+
+    /// Sets the value and marks it as fresh using internal mutability.
+    /// **For internal use only.**
+    fn _set_value(&self, value: PropValue) {
+        *self.value.borrow_mut() = Some(value);
+        self.meta.borrow_mut().status = PropStatus::Fresh;
+        let change_counter = self.meta.borrow().change_counter;
+        // A little bit of safety in case someone wiggles their mouse 4 billion times
+        self.meta.borrow_mut().change_counter = change_counter.wrapping_add(1);
+    }
+
+    /// Sets the prop as stale and clears the cached value using interior mutability.
+    /// **For internal use only.**
+    fn _set_stale(&self) {
```

Comment:
When mark stale, don't remove the value. That way, if the result of a `calculate()` is `NoChange`, you can just mark the cached value as `Fresh` and the old value is restored. We can prevent anyone from accessing the obsolete value as it is private.